components/ContractViewer/index.tsx

```diff
@@ -89,23 +106,35 @@ const ContractViewerInner = () => {
       return
     }
 
-    const addresses = ((router.query.address as string) || '').split(',')
-    for (const addr of addresses) {
-      tryLoadAddress(addr, false)
+    setStatus({
+      status: 'loading',
+    })
+
+    async function restoreSession() {
+      const addresses =
+        typeof router.query.address === 'string'
+          ? router.query.address.split(',')
+          : []
+      const addressPromises = addresses.map((address) => {
+        return tryLoadAddress({
+          address,
+          invalidateRoute: false,
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+          statusCallback: () => {},
+        })
+      })
+      await Promise.all(addressPromises)
+      setStatus({
+        status: 'loaded',
+      })
     }
-    // NOTE: do not add dependencies here or it will cause an infinite loop (idk why)
+
+    void restoreSession()
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [router])
 
   // TODO: fix router to support contract implementations added by user
   // (currently only top-level contracts are supported)
-  useEffect(() => {
```

Comment:
Remove comment above this too which is now not connected to anything

Reply:
Moved the comment to a better place as I still think it's nice to keep around